tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,340 @@
+        keys::check_signing_key_config(repo_info.root_keys)?;
+        root::check_root(root_role_path)?;
+
+        // Create the repo
+        let (s3_stack_arn, bucket_name, bucket_rdn) =
+            create_repo_infrastructure(&mut repo_info).await?;
+        *repo_info.stack_arn = Some(s3_stack_arn);
+        *repo_info.bucket_name = Some(bucket_name.clone());
+        update_root_and_sign_root(&mut repo_info, root_role_path).await?;
+
+        // Upload root.json.
+        info!("Uploading root.json to S3 bucket...");
+        s3::upload_file(
+            &repo_info.s3_region,
+            &bucket_name,
+            &repo_info.prefix,
+            root_role_path,
+        )
+        .await?;
+
+        // Update infra_config with output parameters if not already set
+        if repo_info.metadata_base_url.is_none() {
+            *repo_info.metadata_base_url = Some(
+                Url::parse(
+                    format!(
+                        "https://{}{}/{}/{}/",
+                        &bucket_rdn,
+                        &repo_info.prefix,
+                        shared::getenv("BUILDSYS_VARIANT")?,
+                        shared::getenv("BUILDSYS_ARCH")?
+                    )
+                    .as_str(),
+                )
+                .context(error::ParseUrl { input: &bucket_rdn })?,
+            );
+        }
+        if repo_info.targets_url.is_none() {
+            *repo_info.targets_url = Some(
+                Url::parse(
+                    format!("https://{}{}/targets/", &bucket_rdn, &repo_info.prefix).as_str(),
+                )
+                .context(error::ParseUrl { input: &bucket_rdn })?,
+            );
+        }
+        if repo_info.root_role_url.is_none() {
+            *repo_info.root_role_url = Some(
+                Url::parse(
+                    format!("https://{}{}/root.json", &bucket_rdn, &repo_info.prefix).as_str(),
+                )
+                .context(error::ParseUrl { input: &bucket_rdn })?,
+            );
+        }
+        let root_role_data = fs::read_to_string(&root_role_path).context(error::FileRead {
+            path: root_role_path,
+        })?;
+        let mut d = Sha512::new();
+        d.update(&root_role_data);
+        let digest = hex::encode(d.finalize());
+        repo_config.root_role_sha512 = Some(digest);
+    }
+
+    // Generate Infra.lock
+    info!("Writing Infra.lock...");
+    let yaml_string = serde_yaml::to_string(&infra_config).context(error::InvalidYaml)?;
+    fs::write(
+        toml_path
+            .parent()
+            .context(error::Parent { path: toml_path })?
+            .join("Infra.lock"),
+        yaml_string,
+    )
+    .context(error::FileWrite { path: toml_path })?;
+
+    info!("Complete!");
+    Ok(())
+}
+
+struct ValidRepoInfo<'a> {
+    bucket_name: &'a mut Option<String>,
+    metadata_base_url: &'a mut Option<Url>,
+    prefix: String,
+    pub_key_threshold: &'a String,
+    root_key_threshold: &'a String,
```

Comment:
Parse these in the `new` function to a usize so you don't have to parse it later. Change the function signatures that take threshold as a &str to take it as a `NonZeroUsize`.

```suggestion
    pub_key_threshold: NonZeroUsize,
    root_key_threshold: NonZeroUsize,
```

Reply:
I use thresholds most commonly as strings and only once as a number to check it is within bounds, does it still make sense to parse beforehand? (I'd have to add to_string to the other places I'm using the threshold)